single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/layout/MatrixZoneLayout.java

```diff
@@ -0,0 +1,127 @@
+            for (int col = 0; col < matrix[row].length; col++) {
+                String id = matrix[row][col];
+                SubstationGraph graph = getGraph().getSubstationGraph(id);
+                int hallwayCol = (col + 1) * hallway;
+                if (graph != null) {
+                    double dx = col * maxWidthCol + hallwayCol + diagramPadding.getLeft();
+                    double dy = row * maxHeightRow + hallwayRow + diagramPadding.getTop();
+                    move(graph, dx, dy);
+                }
+                zoneWidth += maxWidthCol + hallway;
+            }
+            zoneHeight += maxHeightRow + hallway;
+        }
+
+        getGraph().setSize(diagramPadding.getLeft() + zoneWidth + diagramPadding.getRight(), zoneHeight + diagramPadding.getBottom());
+    }
+
+    @Override
+    protected List<Point> calculatePolylineSnakeLine(LayoutParameters layoutParam, Pair<Node, Node> nodes,
+                                                     boolean increment) {
+        List<Point> polyline = new ArrayList<>();
+        Node node1 = nodes.getFirst();
+        Node node2 = nodes.getSecond();
+        VoltageLevelGraph vlGraph1 = getGraph().getVoltageLevelGraph(node1);
+        VoltageLevelGraph vlGraph2 = getGraph().getVoltageLevelGraph(node2);
+        SubstationGraph ss1Graph = getGraph().getSubstationGraph(node1).orElse(null);
+        SubstationGraph ss2Graph = getGraph().getSubstationGraph(node2).orElse(null);
+        if (ss1Graph != null && ss1Graph == ss2Graph) { // in the same Substation
+            polyline = layoutBySubstation.get(ss1Graph).calculatePolylineSnakeLine(layoutParam, nodes, increment);
+        } else if (ss1Graph != null && ss2Graph != null &&
+                   model.contains(ss1Graph.getId()) && model.contains(ss2Graph.getId())) { // in the same Zone
+            String ss1Id = ss1Graph.getId();
+            String ss2Id = ss2Graph.getId();
+            Point p1 = vlGraph1.getShiftedPoint(node1);
+            Point p2 = vlGraph2.getShiftedPoint(node2);
+            Direction dNode1 = getNodeDirection(node1, 1);
+            Direction dNode2 = getNodeDirection(node2, 2);
+            // Add starting point
+            polyline.add(p1);
+            // Find snakeline path
+            polyline.addAll(model.buildSnakeline(pathFinder, ss1Id, p1, dNode1, ss2Id, p2, dNode2));
+            // Add ending point
+            polyline.add(p2);
+        } else {
+            // FIXME : substation link but not displayed
+            // not found
+        }
+        return polyline;
+    }
+
+    @Override
+    public void manageSnakeLines(LayoutParameters layoutParameters) {
+        model.computePathFindingGrid(getGraph(), layoutParameters);
+
+        // Draw snakelines between Substations
+        manageSnakeLines(getGraph(), layoutParameters);
+
+        // FIXME: use diagram padding
+        getGraph().setSize(getGraph().getWidth() + 100, getGraph().getHeight() + 100);
```

Comment:
If I get it right, we still have to figure out a way to recompute the graph size now we have drawn the snakelines between substations.

Reply:
Code refactored